File: customaidl/src/main/java/com/xjf/customaidl/AidlService.java
package com.xjf.customaidl;

import android.app.Service;
import android.content.Intent;
import android.os.IBinder;
import android.os.RemoteCallbackList;
import android.os.RemoteException;
import android.util.Log;
import com.xjf.customaidl.aidl.Book;
import com.xjf.customaidl.aidl.BookInterface;

import java.util.ArrayList;
import java.util.List;

/**
 * description: 专家个人信息界面
 *
 * @author xuejingfei
 * create at 19-6-17
 */
public  class AidlService extends Service {
    private final String TAG = AidlService.class.getSimpleName();
    private List<Book> books = new ArrayList<>();

    @Override
    public void onCreate() {
        super.onCreate();
        Log.d(TAG,"onCrate");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        Log.d(TAG,"onStartCommand");
        return Service.START_STICKY;
    }

    @Override
    public void onDestroy(){
        Log.d(TAG,"onDestroy");
    }

    @Override
    public IBinder onBind(Intent intent) {
        return new BookService();
    }

    public class BookService extends BookInterface.Stub{

        @Override
        public void addBook(Book book) throws RemoteException {
            books.add(book);
        }

        @Override
        public List<Book> getBooks() throws RemoteException {
            return books;
        }
    }
}

File: messenager/src/main/java/com/example/messenager/AidlService.kt
package com.example.messenager

import android.app.Service
import android.content.Intent
import android.os.*
import android.util.Log
import com.example.messenager.MessengerActivity.Companion.MSG_FROM_CLIENT

/**
 * Author：xuejingfei
 *
 * Description：
 *
 * Date：2021/6/27 10:44
 */
class AidlService:Service() {

    companion object{
        val TAG = AidlService::class.simpleName
        const val MSG_FROM_SERVICE = 2
    }

    class MessengerHandler:Handler(){
        override fun handleMessage(msg: Message) {
            when(msg.what) {
                 MSG_FROM_CLIENT -> {
                     Log.d(TAG,"receive msg from Client:" + msg.data.getString("msg"))
                     //服务端向客户端发送消息
                     val client = msg.replyTo
                     val messenger = Message.obtain(null,MSG_FROM_SERVICE)
                     val bundle = Bundle()
                     bundle.putString("reply","嗯，你的消息已经收到，稍后会回复您")
                     messenger.data = bundle
                     try {
                         client.send(messenger)
                     }catch (e:RemoteException){
                         e.printStackTrace()
                     }

                 }
                 else ->{
                     super.handleMessage(msg)
                 }
            }
        }
    }


    private val mMessenger:Messenger =  Messenger(MessengerHandler())

    override fun onBind(intent: Intent?): IBinder? {
        return mMessenger.binder
    }

}